### ISSUE TITLE
Meta Continual Learning Scenarios

### ISSUE THREAD

**Issue by hamedhemati** [Continvvm/continuum#136]
Hi and thanks again for the great work!

Similar to the work from Javed et. al [https://arxiv.org/pdf/1905.12588.pdf](https://arxiv.org/pdf/1905.12588.pdf), I want to implement a setup such that dataset classes are split into two individual subsets without any intersection for meta-train and meta-test phases.
I noticed that when creating a scenario, it is not possible to pass a `class_order` list smaller than the actual number of classes, or explicitly use a subset of classes. Therefore I decided to indirectly split the classes via the class order list. Assuming there are in total 100 classes of CIFAR100, as an example I use 80 for meta-train and 20 for meta-test and use a meta task trajectory of length 15 (15 consecutive classes seen at each trajectory). The simplest working solution that I could think of was to split and shuffle the corresponding classes of each phase as below:

```python
from continuum import ClassIncremental
from continuum.datasets import CIFAR100
import random
import copy


cifar100_train = CIFAR100("./data/", download=True)
cifar100_test = CIFAR100("./data/", download=False)

# Initial shuffling of the class order
class_order = list(range(len(cifar100_train.dataset.classes)))
random.shuffle(class_order)

# Meta learning parameters
nb_metatrain_classes = 80 # and thus 20 for met-test
trajectory_length = 15


#=========================== Algorithm
model = X()

# Meta-Train Phase
n_metatrain_trajectories = 50
for meta_task in range(n_trajectories):
    # Shuffle only the classes corresponding to the meta-train phase
    metatrain_classes = copy.copy(class_order[:nb_metatrain_classes])
    random.shuffle(metatrain_classes)
    class_order[:nb_metatrain_classes] = metatrain_classes
    
    # Create a scenarios for the meta-train task and use only the first trajectory_length tasks
    scenario_train = ClassIncremental(cifar100_train, increment=1, class_order=class_order)
    scenario_test = ClassIncremental(cifar100_test, increment=1, class_order=class_order)
    
    meta_train(model, scenario_train[:trajectory_length], scenario_test[:trajectory_length])
    
# Meta-Test Phase
n_metatest_trajectories = 1
for metatest_task in range(n_test_trajectories):
    # Shuffle only the classes corresponding to the meta-test phase
    metatrain_classes = copy.copy(class_order[nb_metatrain_classes:])
    random.shuffle(metatrain_classes)
    class_order[:nb_metatrain_classes] = metatrain_classes
    
    # Create a scenarios for the meta-test task and use only the first trajectory_length tasks
    scenario_train = ClassIncremental(cifar100_train, increment=1, class_order=class_order)
    scenario_test = ClassIncremental(cifar100_test, increment=1, class_order=class_order)

    meta_test(model, scenario_train[:trajectory_length], scenario_test[:trajectory_length])
```

Is there is a simpler/cleaner solution that could possibly be implemented by explicitly separating the dataset classes that I'm not aware of, or would you suggest a similar definition of the scenarios?
Thanks for your time in advance :)

**Comment by arthurdouillard** [comment]
Hey, thanks for the kind words.

I wasn't very aware of this work, but I think I've made something okay-ish:

```python
from continuum import ClassIncremental
from continuum.datasets import CIFAR100, InMemoryDataset
import random
import numpy as np
import copy
from torch.utils.data import DataLoader


cifar100_train = CIFAR100("/local/douillard/", download=True)
cifar100_test = CIFAR100("/local/douillard/", download=False)

# Initial shuffling of the class order
class_order = list(range(len(cifar100_train.dataset.classes)))
random.shuffle(class_order)

# Meta learning parameters
nb_metatrain_classes = 80 # and thus 20 for met-test
trajectory_length = 15

metatrain_classes = np.array(class_order[:nb_metatrain_classes])
print("Train", metatrain_classes)
metatest_classes = np.array(class_order[nb_metatrain_classes:])
print("Test", metatest_classes)

metatrain_train = ClassIncremental(ClassSubset(cifar100_train, metatrain_classes), increment=1)
metatrain_train = ClassIncremental(ClassSubset(cifar100_test, metatrain_classes), increment=1)

metatest_train = ClassIncremental(ClassSubset(cifar100_train, metatest_classes), increment=1)
metatest_train = ClassIncremental(ClassSubset(cifar100_test, metatest_classes), increment=1)

#=========================== Algorithm

# Meta-Train Phase
n_metatrain_trajectories = 50
for meta_task in range(n_metatrain_trajectories):
    # Shuffle only the classes corresponding to the meta-train phase
    # Which is equivalent to shuffling the task ids

    for task_id in np.random.permutation(len(metatrain_train)):
        train_loader = DataLoader(metatrain_train[task_id])
        test_loader = DataLoader(metatrain_train[task_id])
        # Your model learning HERE
        for x, y, t in train_loader:
            # We need to remap y to their original labels
            y = metatrain_classes[y]
    
# Meta-Test Phase
n_metatest_trajectories = 1
for metatest_task in range(n_metatest_trajectories):
    # Shuffle only the classes corresponding to the meta-test phase
    # Which is equivalent to shuffling the task ids

    for task_id in np.random.permutation(len(metatest_train)):
        train_taskset = metatest_train[task_id]
        test_taskset = metatest_train[task_id]
        # Your model learning HERE
```

And with `ClassSubset`:

```python
from continuum.datasets import _ContinuumDataset

class ClassSubset(_ContinuumDataset):
    def __init__(self, dataset, classes):
        self.dataset = dataset
        self.classes = classes
        assert len(self.classes) == len(np.unique(self.classes))
        
    def get_data(self):
        x, y, t = self.dataset.get_data()
        indexes = np.where(np.isin(y, self.classes))[0]
        
        x = x[indexes]
        y = y[indexes]
        if t is not None:
            t = t[indexes]
        
        # We need to remap the class ids so that they are contiguous from O to n
        for i, class_id in enumerate(np.unique(self.classes)):
            y[y == class_id] = i
            
        return x, y, t
```

The only trouble, is that you need to remap the labels in the train loop. But that's fairly fast with vectorization, although it takes a line of code. What do you think of it?

**Comment by TLESORT** [comment]
I have an alternative solution which I think is easier:

````
from continuum import ClassIncremental
from continuum.datasets import CIFAR100
import random
import copy

cifar100_train = CIFAR100("./data/", download=True)
cifar100_test = CIFAR100("./data/", download=False)

scenario_train = ClassIncremental(cifar100_train, increment=1)
scenario_test = ClassIncremental(cifar100_test, increment=1)

# Initial shuffling of the class order
class_order = list(range(len(cifar100_train.dataset.classes)))
random.shuffle(class_order)

# Meta learning parameters
nb_metatrain_classes = 80  # and thus 20 for met-test
trajectory_length = 15

# =========================== Algorithm
model = X()

def meta_train(model, scenario_train, scenario_test, indexes, train=True):
    
    for index in indexes:
        meta_scenario_train = scenario_train[index]
        meta_scenario_test = scenario_test[index]
        
        #[whatever here]

# Meta-Train Phase
n_metatrain_trajectories = 50
for meta_task in range(n_trajectories):
    # Shuffle only the classes corresponding to the meta-train phase
    metatrain_classes = copy.copy(class_order[:nb_metatrain_classes])
    random.shuffle(metatrain_classes)
    class_order[:nb_metatrain_classes] = metatrain_classes
    meta_train(model, scenario_train, scenario_test, class_order, train=True)

# Meta-Test Phase
n_metatest_trajectories = 1
for metatest_task in range(n_test_trajectories):
    # Shuffle only the classes corresponding to the meta-test phase
    metatrain_classes = copy.copy(class_order[nb_metatrain_classes:])
    random.shuffle(metatrain_classes)
    class_order[:nb_metatrain_classes] = metatrain_classes
    meta_train(model, scenario_train, scenario_test, class_order, train=False)
    ```
    
    The idea is, If you create a loop in your meta train function, you can just use the index you need one by one. I hope it can solve what you are trying to do.
    
    An good solution would be use list to slice scenarios but it is not yet supported by continuum scenarios, I will create an issue for it.

**Comment by hamedhemati** [comment]
Thanks, @arthurdouillard, and @TLESORT  for your suggestions. Iterating through sceneario_train and accessing via index seems to be easier without the need to create a new scenario for each meta task. Scenario slicing would be even nicer as it would remove the need to manually access particular classes at each iteration and might also reduce the chances of making mistakes, but probably not urgent.

**Comment by arthurdouillard** [comment]
Yeah you're right. @TLESORT is currently doing a PR (#139) for this. We'll probably release on pypi in the beginning of this week

**Comment by TLESORT** [comment]
The PR has been merged, you can create new scenarios with a original scenario and a list of task ids. It can select only some tasks specifically or change the order of tasks.

You can use it like this for your use case:
```
from continuum.scenarios import create_subscenario
subscenario = create_subscenario(scenario_train, class_order)
for taskset in subscenario:
   [...]
```
I will close the issue, feel free to reopen it if there is a pb :)

**Comment by arthurdouillard** [comment]
Note that you need to install continuum>=1.0.25 to get that feature.
